integration/ruler_test.go

```diff
@@ -179,45 +182,106 @@ func TestRulerEvaluationDelay(t *testing.T) {
 	namespace := "ns"
 	user := "fake"
 
+	evaluationDelay := time.Minute * 5
+
 	configOverrides := map[string]string{
 		"-ruler.storage.local.directory":   filepath.Join(e2e.ContainerSharedDir, "ruler_configs"),
 		"-ruler.poll-interval":             "2s",
 		"-ruler.rule-path":                 filepath.Join(e2e.ContainerSharedDir, "rule_tmp/"),
-		"-ruler.evaluation-delay-duration": "5m", // 5 minutes is clarifying when seeing when a rule is evaluated
+		"-ruler.evaluation-delay-duration": evaluationDelay.String(),
 	}
 
 	// Start Cortex components.
 	require.NoError(t, copyFileToSharedDir(s, "docs/configuration/single-process-config.yaml", cortexConfigFile))
-	require.NoError(t, writeFileToSharedDir(s, filepath.Join("ruler_configs", user, namespace), []byte(cortexRulerEvalTimeConfigYaml)))
+	require.NoError(t, writeFileToSharedDir(s, filepath.Join("ruler_configs", user, namespace), []byte(cortexRulerEvalStaleNanConfigYaml)))
 	cortex := e2ecortex.NewSingleBinaryWithConfigFile("cortex", cortexConfigFile, configOverrides, "", 9009, 9095)
 	require.NoError(t, s.StartAndWaitReady(cortex))
 
 	// Create a client with the ruler address configured
-	c, err := e2ecortex.NewClient("", cortex.HTTPEndpoint(), "", cortex.HTTPEndpoint(), "")
+	c, err := e2ecortex.NewClient(cortex.HTTPEndpoint(), cortex.HTTPEndpoint(), "", cortex.HTTPEndpoint(), "")
 	require.NoError(t, err)
 
-	// Wait until the rule is evaluated
-	require.NoError(t, cortex.WaitSumMetrics(e2e.Greater(0), "cortex_prometheus_rule_evaluations_total"))
-
 	now := time.Now()
 
-	result, err := c.QueryRange("time_eval", now.Add(-10*time.Minute), now, time.Minute)
+	// Generate series that includes stale nans
+	var samplesToSend int = 10
+	series := prompb.TimeSeries{
+		Labels: []prompb.Label{
+			{Name: "__name__", Value: "a_sometimes_stale_nan_series"},
+			{Name: "instance", Value: "sometimes-stale"},
+		},
+	}
+	series.Samples = make([]prompb.Sample, samplesToSend)
+	posStale := map[int]struct{}{
+		2: {},
+	}
+
+	// Create samples, that are delayed by the evaluation delay with increasing values.
+	for pos := range series.Samples {
+		series.Samples[pos].Timestamp = e2e.TimeToMilliseconds(now.Add(-evaluationDelay)) + 1000*int64(pos)
+		series.Samples[pos].Value = float64(pos + 1)
+
+		// insert staleness marker at the positions marked by posStale
+		if _, stale := posStale[pos]; stale {
+			series.Samples[pos].Value = math.Float64frombits(value.StaleNaN)
+		}
+	}
+
+	// Insert metrics
+	res, err := c.Push([]prompb.TimeSeries{series})
 	require.NoError(t, err)
-	require.Equal(t, model.ValMatrix, result.Type())
+	require.Equal(t, 200, res.StatusCode)
 
-	// Iterate through the values recorded and ensure they exist in the past.
-	matrix := result.(model.Matrix)
+	// Wait until the rule is evaluated for the first time
+	require.NoError(t, cortex.WaitSumMetrics(e2e.Greater(1.0), "cortex_prometheus_rule_evaluations_total"))
```

Comment:
Since it's greater than 1 it will wait for 2 evaluations. Is that expected or you wanted `e2e.Greater(0)`?

Reply:
Also, I believe this test can be flaky because of this check. What you want is to wait 1 more evaluation after the `Push()`, so ideally you should get the the current `cortex_prometheus_rule_evaluations_total` and `WaitSumMetric()` by that value + 1.

Reply:
Good point, I have implemented as suggest, `Push -> SumMetric -> WaitSumMetric(Greater(result))`